### PR TITLE
chore(k8sclusterreceiver): 🔥  remove deprecated kubernetes API resources

### PR DESCRIPTION
**Pull request by prashant-shahi**
**Description:**
Remove support for deprecated Kubernetes API resources: 
- `batch/v1beta1`
- `autoscaling/v2beta2`

This also resolves the issue with double reporting metrics for hpa and cronjob in certain k8s versions.

**Link to tracking Issue(s):**
#23612 #26551

**Testing:**
From source, built the custom image [coolboi567/otelcontribcol:0.83.1](https://hub.docker.com/layers/coolboi567/otelcontribcol/0.83.1/images/sha256-fe111e1dff87a26eb64217a8505b84679263c8d7b3ffa16656bba1c1865052d5?context=explore) and tested in kubernetes cluster `v1.25` as well as `v1.27`.

In K8s `v1.25`, we no longer see the warnings in the logs about usage of deprecated APIs.

**Documentation:**
N/A

### CODE REVIEW

File: receiver/k8sclusterreceiver/README.md [hunk 1]
```diff
@@ -67,6 +67,9 @@ The following allocatable resource types are available.
   - storage
 - `metrics`: Allows to enable/disable metrics.
 - `resource_attributes`: Allows to enable/disable resource attributes.
+- `ignore_deprecated_resource` (default = `false`): Whether to ignore deprecated Kubernetes API resources. If set to true,
```
Comment (dmitryax) [inline receiver/k8sclusterreceiver/README.md:70]: `ignore_deprecated_resources` would probably be a better name? (I know I mentioned `ignore_deprecated_resource` in the issue :))

**Review by dmitryax — COMMENTED** [review]
@prashant-shahi thank you for the PR. Can you please add a changelog entry?

**Comment by jinja2** [comment]
Can we drop support for the hpa and cronjob beta apis? I am not sure if we have a compatibility matrix with k8s versions. Hpa was promoted to stable in v1.23 and cronjob in v1.21. Upstream k8s currently maintains 1.25 to 1.28.

Below are EOL dates for the major managed clusters, all of which have no support for 1.22 (the last version which would need the hpa beta):

https://endoflife.date/google-kubernetes-engine
https://endoflife.date/azure-kubernetes-service
https://endoflife.date/amazon-eks

I think if we can commit to some k8s version compatibility window, it would make maintenance of k8s receivers simpler. A good place to start might be to keep up with the major managed cluster providers.

**Comment by dmitryax** [comment]
@jinja2, that makes sense. It depends on the definition of "support" from the collector side. Should we support collecting metrics from the deprecated resource versions? I believe we should. In that case, dropping support them now makes the supported window pretty narrow: 1.28, 1.27, 1.26 due to HPA beta removal in 1.26. I'm in favor of introducing this config option to provide a smoother transition. This will likely be needed going forward as well.

File: receiver/k8sclusterreceiver/README.md (hunk 1 above)
Comment (prashant-shahi) [reply receiver/k8sclusterreceiver/README.md]: agreed. That looks like better name.

Updated.

**Comment by prashant-shahi** [comment]
> Can you please add a changelog entry?

Changelog has been updated.

> dropping them now makes the supported window pretty narrow: 1.28, 1.27, 1.26 due to HPA beta removal in 1.26. I'm in favor of introducing this config option to provide a smoother transition.

Absolutely. Introducing this option can also help the folks who do not want to upgrade to newest K8s version but want to get rid of the deprecation warning.

Also, the usage of the deprecated K8s API causes the failure of managed clusters to auto-upgrade to newer k8s version (as reported in GKE and OpenShift).

**Comment by jinja2** [comment]
I have started a new [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26551) which provides more context on api versioning in k8s. The tldr is that we can use the `v2` api for hpas which we created in the `v2beta2` version format, starting from k8s v1.23. I don't think we should add this flag. Instead we should decide on our k8s version compatibility. Removing `v2beta2` here still makes us compatible with `latest - 5` versions of k8s which to me seems like a very reasonable support duration.

**Comment by prashant-shahi** [comment]
@jinja2 Thanks for the detailed issue on the behaviour. It was really informative.

Updating the PR to remove the deprecated `v2beta2` APIs instead of adding the flag.

File: CHANGELOG.md [hunk 2]
```diff
@@ -55,6 +55,7 @@ If you are looking for developer-facing changes, check out [CHANGELOG-API.md](./
 - `snmpreceiver`: Timeout for SNMP requests can now be configured. (#25885)
 - `telemetrygen`: The telemetrygen now supports setting the log's body (#26031)
 - `awsxrayexporter`: add `exporter.awsxray.skiptimestampvalidation` Alpha feature gate to remove xray timestamp restriction on first 32 bits of trace id (#26041)
+- `k8sclusterreceiver`: Remove deprecated Kubernetes API resources (#26516)
```
Comment (dmitryax) [inline CHANGELOG.md:58]: This file is autogenerated. Can you please add a file in https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/.chloggen instead?

**Review by dmitryax — COMMENTED** [review]
LGTM other than the changelog issue

**Review by dmitryax — APPROVED** [review]
Thank you, @prashant-shahi!

**Review by TylerHelmuth — COMMENTED** [review]
Based on this issue, we should update the README that this breaks <= 1.22 users right?

File: CHANGELOG.md (hunk 2 above)
Comment (prashant-shahi) [reply CHANGELOG.md]: My bad.

It is updated now.

**Comment by prashant-shahi** [comment]
> Based on this issue, we should update the README that this breaks <= 1.22 users right?

Include that in changelog? Yeah, sure.

However, I am not sure if we need to include it in README.
It's been around a year since the end of life for `1.22` support.

Ref: https://endoflife.date/kubernetes

**Comment by TylerHelmuth** [comment]
@prashant-shahi if a user enabled this component in a k8s 1.22 cluster with the default values would the collector crash?  If so I think we should call that out in the README since there are definitely still people using <= 1.22.

**Comment by jinja2** [comment]
@TylerHelmuth it won't crash, the receiver will start up ok but HPA metrics will stop. I think a user-facing changelog entry to call out dropped support for the api is good enough.

File: .chloggen/k8sclusterreceiver-remove-deprecated-resources.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Remove deprecated Kubernetes API resources"
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [23612,26551]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (dmitryax) [inline .chloggen/k8sclusterreceiver-remove-deprecated-resources.yaml:18]:
```suggestion
subtext: |
  Drop support of v2beta2 version of HorizontalPodAutoscaler and v1beta1 version of CronJob resources. 
  Note that metrics for those resources will not be emitted anymore on Kubernetes 1.22 and older.
```

File: .chloggen/k8sclusterreceiver-remove-deprecated-resources.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: "deprecation"
```
Comment (dmitryax) [inline .chloggen/k8sclusterreceiver-remove-deprecated-resources.yaml:4]:
```suggestion
change_type: "breaking"
```

**Review by TylerHelmuth — APPROVED** [review]

**Comment by prashant-shahi** [comment]
@dmitryax @jinja2 @TylerHelmuth Thanks for the thorough review and suggestions in the PR.

Can the related issues be closed now or after the next release?

**Comment by TylerHelmuth** [comment]
@prashant-shahi you can close it now

**Comment by prashant-shahi** [comment]
> you can close it now

@TylerHelmuth I am unable to close them. Since the related issues were created by other folks.

**Pull request merged**